stl/inc/flat_map

```diff
@@ -0,0 +1,455 @@
+            _Key_it += _Off;
+            _Mapped_it += _Off;
+            return *this;
+        }
+
+        type& operator-=(difference_type _Off) {
+            _Key_it -= _Off;
+            _Mapped_it -= _Off;
+            return *this;
+        }
+
+        type operator+(difference_type _Off) const {
+            type _Tmp = *this;
+            return _Tmp += _Off;
+        }
+
+        type operator-(difference_type _Off) const {
+            type _Tmp = *this;
+            return _Tmp -= _Off;
+        }
+
+        reference operator[](difference_type _Off) const {
+            return *(*this + _Off);
+        }
+
+        difference_type operator-(const type& _Right) const {
+            return _Key_it - _Right._Key_it;
+        }
+
+        friend type operator+(difference_type _Off, const type& _Right);
+
+    private:
+        _Key_iterator_t _Key_it;
+        _Mapped_iterator_t _Mapped_it;
+    };
+
+    static_assert(_STD swappable<type>);
+};
+
+template <_STD random_access_iterator _Key_iterator_t, _STD random_access_iterator _Mapped_iterator_t>
+_EXPORT_STD typename _Flat_map_iterator_Impl<_Key_iterator_t, _Mapped_iterator_t>::type operator+(
+    typename _Flat_map_iterator_Impl<_Key_iterator_t, _Mapped_iterator_t>::type::difference_type _Off,
+    const typename _Flat_map_iterator_Impl<_Key_iterator_t, _Mapped_iterator_t>::type& _Right) {
+    return _Right + _Off;
+}
+
+_EXPORT_STD
+template <class Key, class T, class Compare, class KeyContainer, class MappedContainer>
+    requires same_as<Key, typename KeyContainer::value_type> && same_as<T, typename MappedContainer::value_type>
+class flat_map {
+public:
+    using key_type               = Key;
+    using mapped_type            = T;
+    using value_type             = pair<key_type, mapped_type>;
+    using key_compare            = Compare;
+    using reference              = pair<const key_type&, mapped_type&>;
+    using const_reference        = pair<const key_type&, const mapped_type&>;
+    using size_type              = size_t;
+    using difference_type        = ptrdiff_t;
+    using key_container_type     = KeyContainer;
+    using mapped_container_type  = MappedContainer;
+    using iterator               = _STD _Flat_map_iterator_Impl<typename key_container_type::const_iterator,
+        typename mapped_container_type::iterator>::type;
+    using const_iterator         = _STD _Flat_map_iterator_Impl<typename key_container_type::const_iterator,
+        typename mapped_container_type::const_iterator>::type;
+    using reverse_iterator       = _STD reverse_iterator<iterator>;
+    using const_reverse_iterator = _STD reverse_iterator<const_iterator>;
+
+    static_assert(_STD random_access_iterator<iterator>);
+
+    using value_compare = typename _STD _Flat_value_compare<key_type, mapped_type, key_compare>::value_compare;
+    using containers    = typename _STD _Flat_Container<key_container_type, mapped_container_type>::container;

```

Comment:
By the way, as a followup to https://github.com/microsoft/STL/pull/3894, we need to remove these `typename`s finally.

Reply:
Removed.